libraries/AC_Autorotation/AC_Autorotation.cpp

```diff
@@ -251,54 +321,121 @@ float AC_Autorotation::get_rpm(bool update_counter)
+    float b = _param_solidity*6.28f;
+    _disc_area=3.14*sq(_param_diameter*0.5f);
+    float lambda = (-(b/8.0f) + safe_sqrt((sq(b))/64.0f +((b/3.0f)*_col_hover_rad)))*0.5f;
+    float freq=_governed_rpm/60.0f;
+    float tip_speed= freq*6.28f*_param_diameter*0.5f;
+    _lift_hover = ((1.225f*sq(tip_speed)*(_param_solidity*_disc_area))*((_col_hover_rad/3.0f) - (lambda/2.0f))*5.8f)*0.5f;
+
+    //estimate rate of descent
+    float omega_auto=(_param_head_speed_set_point/60.0f)*6.28f;
+    float tip_speed_auto = omega_auto*_param_diameter*0.5f;
+    float c_t = _lift_hover/(0.6125f*_disc_area*sq(tip_speed));
+    _est_rod=((0.25f*(_param_solidity*0.011f/c_t)*tip_speed_auto)+((sq(c_t)/(_param_solidity*0.011f))*tip_speed_auto));
+
+    //estimate rotor C_d
+    _c=(_lift_hover/(sq(_est_rod)*0.5f*1.225f*_disc_area))*1.15f;
+    _c=constrain_float(_c, 0.7f, 1.4f);
+
+    //calc flare altitude
+    float des_spd_fwd=_param_target_speed*0.01f;
+    calc_flare_alt(-_est_rod,des_spd_fwd);
+
+    //initialize sink rate monitor and flare bools
+    _flare_complete = false;
+    _flare_update_check = false;
+    _avg_sink_deriv = 0.0f;
+    _avg_sink_deriv_sum = 0.0f;
+    _index_sink_rate = 0;
+    memset(_curr_sink_deriv, 0, sizeof(_curr_sink_deriv));
+
+    gcs().send_text(MAV_SEVERITY_INFO, "Ct/sigma=%f W=%f kg flare_alt=%f C=%f", c_t/_param_solidity, _lift_hover/9.8065f, _flare_alt_calc*0.01f, _c);
+
+}
+
+
+
+void AC_Autorotation::calc_flare_alt(float sink_rate, float fwd_speed)
+{
+    //compute speed module and glide path angle during descent
+    float speed_module=norm(sink_rate,fwd_speed);
+    float glide_angle=safe_asin(3.14/2-(fwd_speed/speed_module));
+
+    //estimate inflow velocity at beginning of flare
+    float inflow= -speed_module*sinf(glide_angle+radians(AP_ALPHA_TPP));
+
+    //estimate flare duration
+    float k_1=fabsf((-sink_rate+0.001f-safe_sqrt(_lift_hover/_c))/(-sink_rate+0.001f+safe_sqrt(_lift_hover/_c)));
+    float k_2=fabsf((inflow-safe_sqrt(_lift_hover/_c))/(inflow+safe_sqrt(_lift_hover/_c)));
+    float delta_t_flare=(0.5f*(_lift_hover/(9.8065*_c))*safe_sqrt(_c/_lift_hover)*logf(k_1))-(0.5f*(_lift_hover/(9.8065*_c))*safe_sqrt(_c/_lift_hover)*logf(k_2));
+
+    //estimate flare delta altitude
+    float a=(2*safe_sqrt(_c*9.8065/(_lift_hover/9.8065)))*delta_t_flare + (2*safe_sqrt(_c*9.8065/(_lift_hover/9.8065)))*(0.5f*(_lift_hover/(9.8065*_c))*safe_sqrt(_c/_lift_hover)*logf(k_1));
+    float x=1-expf(a);
+    float s=1-expf((2*safe_sqrt(_c*9.8065/(_lift_hover/9.8065)))*(0.5f*(_lift_hover/(9.8065*_c))*safe_sqrt(_c/_lift_hover)*logf(k_1)));
+    float d=safe_sqrt(_lift_hover/_c);
+    float flare_distance=((2*d/(2*safe_sqrt(_c*9.8065/(_lift_hover/9.8065))))*(a-logf(fabsf(x))-(2*safe_sqrt(_c*9.8065/(_lift_hover/9.8065)))*(0.5f*(_lift_hover/(9.8065*_c))*safe_sqrt(_c/_lift_hover)*logf(k_1)) +logf(fabsf(s))))-d*delta_t_flare;
+    float delta_h= -flare_distance*cosf(radians(AP_ALPHA_TPP));
```

Comment:
Lots of `GRAVITY_MSS` need here. `_c*9.8065/(_lift_hover/9.8065)` turns up lots which is `c*GRAVITY_MSS*GRAVITY_MSS/_lift_hover`. You could break out G^2 as a const.

Reply:
done